Purge inactive friends and avoids

jauggy opened the pull request:
The purpose of this is to help fix Kobold's error https://ptb.discord.com/channels/549281623154229250/1310908634036703302
This user has 1000 avoids/blocks/ignores. It is basically difficult for him to trim that list down and it causing issues.

![Screenshot 2024-12-11 at 3 11 04 pm](https://github.com/user-attachments/assets/96a2cb5a-9d47-4df8-8714-edabbeb966dc)

![Screenshot 2024-12-11 at 3 12 42 pm](https://github.com/user-attachments/assets/ed2bd120-0b31-44ab-a397-f2f861317cb3)

jauggy commented:
@L-e-x-o-n whenever you have to time can you please put this on integration server. I would like to test.

jauggy commented on lib/teiserver/account/libs/relationship_lib.ex:
```diff
@@ -388,4 +388,25 @@ defmodule Teiserver.Account.RelationshipLib do
+        days_not_logged_in
+      ])
+
+    results.num_rows
```
Any sample code for this? I have no idea how to do this.

geekingfrog replied:
no need for a friend request, you can directly use [create_friend](https://github.com/geekingfrog/teiserver/blob/d7df6d3d168415d4932b4b6605063743e48be969/lib/teiserver/account.ex#L1680-L1687) since you only care about the db relationship, not the flow leading to create this relationship.

And to create the user, that's the idea. I suggest you locate/create a function like `user_attrs` that will return the plain map and put that in the fixture folder in tests. This way, you can create a default map, modify it for the purpose of you tests, and then call `create_user` on it. I'm using this pattern a bit for testing oauth stuff: https://github.com/geekingfrog/teiserver/blob/d7df6d3d168415d4932b4b6605063743e48be969/test/teiserver/o_auth/gen_token_test.exs#L13

jauggy commented on lib/teiserver_web/live/account/relationship/index.ex:
```diff
@@ -255,6 +265,58 @@ defmodule TeiserverWeb.Account.RelationshipLive.Index do
+      |> length()
+
+    socket =
+      socket |> assign(:purge_friends_message, "#{num_friends_deleted} inactive friends purged.")
```
Tested with
```
      |> put_flash(:info, "Test")
```
Nothing showed up

geekingfrog replied:
ah damn, this is because this is a liveview, and the flash feature is only in place for regular views :(
it would be nice to enable a flash feature for live views, but that's out of scope for this change.

jauggy commented on lib/teiserver_web/live/account/relationship/index.ex:
```diff
@@ -413,4 +475,21 @@ defmodule TeiserverWeb.Account.RelationshipLive.Index do
+        number * 365.0 / 12
+
+      true ->
+        raise("Incorrect value assigned to :purge_cutoff")
```
The error should be received by the dev. The user can only choose options so would never get an error unless a future dev added an option that didn't work. So I probably need to make a test function to make sure a future dev doesn't input an option that would break.

geekingfrog replied:
No, this is not really a dev here, because this value is coming from a user. Sure, using the web page as intended will not give you weird values but anyone could submit whatever they want.

jauggy replied:
How could a user submit a value other than the options I have given them?

jauggy commented:
Updated based on @geekingfrog feedback. I didn't add any test for friend purge because I am using existing function made by Teifion (I just loop through them). Added test for relationship avoids and other tests.

L-e-x-o-n commented:
@jauggy deployed

I think the idea is nice but I would be a bit worried that I might delete some friends I care about when using this feature, especially when used with a more recent timeframes... A confirmation popup would be nice, it would be even nicer if that popup could show the number and a list of users that would be purged (before clicking confirm).

jauggy commented:
I can remove the 1 month option. I will see if I can get a count first.

Do we have any existing pages with popups? I have no idea how they work.

jauggy commented:
Added a confirm dialog for purging friends. Not sure how to show dynamic information in the confirmation.
Removed 1 month as an option.

jauggy commented:
@L-e-x-o-n whenever you have time I'd like to request this be placed on integration server again. Last time I got kobold to test and it completely wiped his avoid list. This occured probably because there's something wrong with last_login column. It's possibly always nil. I adjusted code to make it check `last_login_timex` instead. But I think last_login being always nil is problematic and can be looked at in a future ticket.

L-e-x-o-n commented:
> @L-e-x-o-n whenever you have time I'd like to request this be placed on integration server again. Last time I got kobold to test and it completely wiped his avoid list. https://discord.com/channels/549281623154229250/1310908634036703302/1318145574918291536
> 
> This occured probably because there's something wrong with last_login column. It's possibly always nil. I adjusted code to make it check `last_login_timex` instead. But I think last_login being always nil is problematic and can be looked at in a future ticket.
> 
> I have no idea why Teifion made both these columns

Deployed again

jauggy commented:
@L-e-x-o-n had to do a few more changes. When you have a chance please deploy again to integration. Thanks.

L-e-x-o-n commented:
> @L-e-x-o-n had to do a few more changes. When you have a chance please deploy again to integration. Thanks.

Done
